File: src/app/www/controllers/validate.rs
use std::sync::Arc;

use axum::{
    extract::{Form, State},
    http::HeaderMap,
};
use maud::{html, Markup};

use crate::{
    app::www::{
        components::{notification, submit},
        fragments::{
            company_field, company_xp_field, compensation_field, email_field, level_field,
            location_field, title_field, total_xp_field,
        },
        i18n::I18n,
        models::{HXTriggerName, ParsedForm, UnparsedForm, ValidatedForm},
    },
    domain::use_cases,
    infra::{CompanyRepository, LocationRepository, TitleRepository},
};

pub async fn post(
    headers: HeaderMap,
    State(company_repo): State<Arc<dyn CompanyRepository>>,
    State(location_repo): State<Arc<dyn LocationRepository>>,
    State(title_repo): State<Arc<dyn TitleRepository>>,
    Form(unparsed_form): Form<UnparsedForm>,
) -> Markup {
    let hx_trigger_name = match HXTriggerName::try_from(headers) {
        Ok(hx_trigger_name) => hx_trigger_name,
        _ => return error(),
    };

    let parsed_form = ParsedForm::from(unparsed_form);
    let disabled = ValidatedForm::try_from(parsed_form.clone()).is_err();

    let field = match (&hx_trigger_name).into() {
        "email" => email_field::view(parsed_form.email),
        "company" => match use_cases::fetch_companies(company_repo).await {
            Ok(companies) => company_field::view(parsed_form.company, companies),
            Err(use_cases::fetch_companies::Error::Unknown(_)) => return error(),
        },
        "title" => match use_cases::fetch_titles(title_repo).await {
            Ok(titles) => title_field::view(parsed_form.title, titles),
            Err(use_cases::fetch_titles::Error::Unknown(_)) => return error(),
        },
        "level" => level_field::view(parsed_form.level),
        "location" => match use_cases::fetch_locations(location_repo).await {
            Ok(locations) => location_field::view(parsed_form.location, locations),
            Err(use_cases::fetch_locations::Error::Unknown(_)) => return error(),
        },
        "compensation" => compensation_field::view(parsed_form.compensation),
        "company_xp" => company_xp_field::view(parsed_form.company_xp),
        "total_xp" => total_xp_field::view(parsed_form.total_xp),
        _ => html! {},
    };

    html! {
        (field)
        (submit::view(disabled, I18n::Send.translate()))
    }
}

fn error() -> Markup {
    html! { (notification::view(Some(I18n::ValidationError.translate()))) }
}

File: src/app/www/controllers/validate.ts
import { Context } from "hono";
import { CompanyRepository, LocationRepository, TitleRepository } from "@infra";
import { parseForm } from "./utils/mod.ts";
import { ParsedForm, ValidatedForm, formSchema } from "../models/mod.ts";
import { Maybe, Result } from "@utils";
import { Notification, Submit } from "../components/mod.ts";
import { I18n } from "../i18n.ts";
import { CompanyField, CompanyXpField, CompensationField, EmailField, LevelField, LocationField, RemoteField, TitleField, TotalXpField } from "../fragments/mod.ts";
import { fetchCompanies, fetchLocations, fetchTitles } from "@domain";

export async function post(
    c: Context,
    companyRepo: CompanyRepository,
    locationRepo: LocationRepository,
    titleRepo: TitleRepository
) {
    const name = c.req.header("HX-Trigger-Name");

    const unparsedForm = await parseForm(c, formSchema);
    const parsedForm = ParsedForm.fromUnparsedForm(unparsedForm);
    const disabled = Result.isErr(ValidatedForm.tryFromParsedForm(parsedForm));

    const field = await generateField(
        name,
        parsedForm,
        companyRepo,
        locationRepo,
        titleRepo
    );

    return c.html([field, Submit({
        disabled,
        label: I18n.translate("send")
    })].join(""));
}

async function generateField(
    name: string | undefined,
    parsedForm: ParsedForm,
    companyRepo: CompanyRepository,
    locationRepo: LocationRepository,
    titleRepo: TitleRepository
) {
    switch (name) {
        case "email": return EmailField({ internals: parsedForm.email });
        case "company": return Result.match(await fetchCompanies(companyRepo), {
            onOk: (companies) => CompanyField({ internals: parsedForm.company, companies }),
            onErr: () => buildErrorNotification()
        });
        case "title": return Result.match(await fetchTitles(titleRepo), {
            onOk: (titles) => TitleField({ internals: parsedForm.title, titles }),
            onErr: () => buildErrorNotification()
        });
        case "level": return LevelField({ internals: parsedForm.level });
        case "location": return Result.match(await fetchLocations(locationRepo), {
            onOk: (locations) => LocationField({ internals: parsedForm.location, locations }),
            onErr: () => buildErrorNotification()
        });
        case "compensation": return CompensationField({ internals: parsedForm.compensation });
        case "companyXp": return CompanyXpField({ internals: parsedForm.companyXp });
        case "totalXp": return TotalXpField({ internals: parsedForm.totalXp });
        case "remoteVariant":
        case "remoteDayCount":
        case "remoteBase":
        case "remoteLocation": return RemoteField({ internals: parsedForm.remote });
        default: return "";
    }
}

function buildErrorNotification() {
    return Notification({
        notification: Maybe.some(I18n.translate("validation_error"))
    });
}